dbms/src/Functions/ExtractString.h

```diff
@@ -0,0 +1,187 @@
+#include <Common/UTF8Helpers.h>
+
+#include <algorithm>
+#include <climits>
+#include <cstring>
+#include <memory>
+#include <utility>
+
+#ifdef __SSE4_2__
+#    include <nmmintrin.h>
+#endif
+
+namespace DB
+{
+// used by FunctionsStringSimilarity and FunctionsStringHash
+// includes extracting ASCII ngram, UTF8 ngram, ASCII word and UTF8 word
+template <size_t N, bool CaseInsensitive>
+struct ExtractStringImpl
+{
+    static constexpr size_t default_padding = 16;
+
+    static ALWAYS_INLINE size_t readASCIICodePoints(UInt8 * code_points, const char *& pos, const char * end)
+    {
+        /// Offset before which we copy some data.
+        constexpr size_t padding_offset = default_padding - N + 1;
+        /// We have an array like this for ASCII (N == 4, other cases are similar)
+        /// |a0|a1|a2|a3|a4|a5|a6|a7|a8|a9|a10|a11|a12|a13|a14|a15|a16|a17|a18|
+        /// And we copy                                ^^^^^^^^^^^^^^^ these bytes to the start
+        /// Actually it is enough to copy 3 bytes, but memcpy for 4 bytes translates into 1 instruction
+        memcpy(code_points, code_points + padding_offset, roundUpToPowerOfTwoOrZero(N - 1) * sizeof(UInt8));
+        /// Now we have an array
+        /// |a13|a14|a15|a16|a4|a5|a6|a7|a8|a9|a10|a11|a12|a13|a14|a15|a16|a17|a18|
+        ///              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        /// Doing unaligned read of 16 bytes and copy them like above
+        /// 16 is also chosen to do two `movups`.
+        /// Such copying allow us to have 3 codepoints from the previous read to produce the 4-grams with them.
+        memcpy(code_points + (N - 1), pos, default_padding * sizeof(UInt8));
+
+        if constexpr (CaseInsensitive)
+        {
+            /// We really need template lambdas with C++20 to do it inline
+            unrollLowering<N - 1>(code_points, std::make_index_sequence<padding_offset>());
+        }
+        pos += padding_offset;
+        if (pos > end)
+            return default_padding - (pos - end);
+        return default_padding;
+    }
+
+    // used by FunctionsStringHash
+    // it's not easy to add padding for ColumnString, so we need safety check each memcpy
+    static ALWAYS_INLINE size_t readASCIICodePointsNoPadding(UInt8 * code_points, const char *& pos, const char * end)
```

Comment:
We can avoid this function (and use the same as above) if we will pass internal ColumnString of ColumnConst instead of std::string.